### PR TITLE
MTDSA-20185 Migrate endpoints from IIR monolith

### PR DESCRIPTION
**Pull request by mattstephens-hmrc**

### CODE REVIEW

File: resources/public/api/conf/1.0/application.yaml [hunk 1]
```diff
@@ -18,32 +41,28 @@ components:
       schema:
         type: string
         enum: [
-          "application/vnd.hmrc.1.0+json",
-          "application/vnd.hmrc.1.0+xml"
+            "application/vnd.hmrc.1.0+json",
```
Comment (SamuelJacobsHMRC) [inline resources/public/api/conf/1.0/application.yaml:44]: Usually in our application.yaml files under components we have the `securitySchemes`. The headers are normally specified under each individual endpoint yaml file so probably dont need this section

File: test/shared/support/DownstreamResponseMappingSupportSpec.scala [hunk 2]
```diff
@@ -82,7 +82,7 @@ class DownstreamResponseMappingSupportSpec extends UnitSpec {
         "use the mapping and wrap with main error type of BadRequest" in {
           mapping.mapDownstreamErrors(errorCodeMap)(
             ResponseWrapper(correlationId, DownstreamErrors(List(DownstreamErrorCode("ERR1"), DownstreamErrorCode("ERR2"))))) shouldBe
-            errors.ErrorWrapper(correlationId, BadRequestError, Some(Seq(Error1, Error2)))
+            errors.ErrorWrapper(correlationId, BadRequestError, Some(List(Error1, Error2)))
```
Comment (SamuelJacobsHMRC) [inline test/shared/support/DownstreamResponseMappingSupportSpec.scala:85]:
```suggestion
           ErrorWrapper(correlationId, BadRequestError, Some(List(Error1, Error2)))
```

File: test/shared/support/DownstreamResponseMappingSupportSpec.scala [hunk 3]
```diff
@@ -112,8 +112,8 @@ class DownstreamResponseMappingSupportSpec extends UnitSpec {
 
     "the error code is an OutboundError with multiple errors" must {
       "return the error as is (in an ErrorWrapper)" in {
-        mapping.mapDownstreamErrors(errorCodeMap)(ResponseWrapper(correlationId, errors.OutboundError(ErrorBvrMain, Some(Seq(ErrorBvr))))) shouldBe
-          errors.ErrorWrapper(correlationId, ErrorBvrMain, Some(Seq(ErrorBvr)))
+        mapping.mapDownstreamErrors(errorCodeMap)(ResponseWrapper(correlationId, errors.OutboundError(ErrorBvrMain, Some(List(ErrorBvr))))) shouldBe
```
Comment (SamuelJacobsHMRC) [inline test/shared/support/DownstreamResponseMappingSupportSpec.scala:115]:
```suggestion
        mapping.mapDownstreamErrors(errorCodeMap)(ResponseWrapper(correlationId,OutboundError(ErrorBvrMain, Some(List(ErrorBvr))))) shouldBe
```

File: test/v1/connectors/CreateAmendForeignConnectorSpec.scala [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        def taxYear: TaxYear = TaxYear.fromMtd("2019-20")
+
+        val outcome = Right(ResponseWrapper(correlationId, ()))
+
+        willPut(
+          url = s"$baseUrl/income-tax/income/foreign/$nino/${taxYear.asMtd}",
+          body = amendForeignRequestBody
+        ).returns(Future.successful(outcome))
+
+        await(connector.amendForeign(amendForeignRequest)) shouldBe outcome
```
Comment (SamuelJacobsHMRC) [inline test/v1/connectors/CreateAmendForeignConnectorSpec.scala:43]:
```suggestion
        val result = await(connector.amendForeign(amendForeignRequest))
        result shouldBe outcome
```

etc.

File: test/v1/connectors/MockCreateAmendForeignConnector.scala [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+import uk.gov.hmrc.http.HeaderCarrier
+import v1.models.request.createAmend.CreateAmendForeignRequest
+
+import scala.concurrent.{ExecutionContext, Future}
+
+trait MockCreateAmendForeignConnector extends MockFactory {
+
+  val mockAmendForeignConnector: CreateAmendForeignConnector = mock[CreateAmendForeignConnector]
+
+  object MockAmendForeignConnector {
```
Comment (SamuelJacobsHMRC) [inline test/v1/connectors/MockCreateAmendForeignConnector.scala:31]:
```suggestion
  object MockedCreateAmendForeignConnector {
```

(now is probably a good opportunity to get the things that should be called `Mocked...` to be so)

File: test/v1/connectors/MockCreateAmendForeignConnector.scala [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+import org.scalamock.scalatest.MockFactory
+import shared.connectors.DownstreamOutcome
+import uk.gov.hmrc.http.HeaderCarrier
+import v1.models.request.createAmend.CreateAmendForeignRequest
+
+import scala.concurrent.{ExecutionContext, Future}
+
+trait MockCreateAmendForeignConnector extends MockFactory {
+
+  val mockAmendForeignConnector: CreateAmendForeignConnector = mock[CreateAmendForeignConnector]
```
Comment (SamuelJacobsHMRC) [inline test/v1/connectors/MockCreateAmendForeignConnector.scala:29]:
```suggestion
  val mockCreateAmendForeignConnector: CreateAmendForeignConnector = mock[CreateAmendForeignConnector]
```

File: test/v1/services/CreateAmendForeignServiceSpec.scala [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+    CreateAmendForeignRequest(nino = Nino(nino), taxYear = TaxYear.fromMtd(taxYear), body = amendForeignRequestBody)
+
+  trait Test extends MockCreateAmendForeignConnector {
+    implicit val logContext: EndpointLogContext = EndpointLogContext("c", "ep")
+
+    val service: CreateAmendForeignService = new CreateAmendForeignService(
+      connector = mockAmendForeignConnector
+    )
+
+  }
```
Comment (SamuelJacobsHMRC) [inline test/v1/services/CreateAmendForeignServiceSpec.scala:67]: Don't think this Test trait is necessary at all. The CreaetAmendForeignServiceSpec could extend the MockCreateAmendForeignConnector and then the service just be created as a private val. The logContext can be removed entirely

File: test/v1/services/MockCreateAmendForeignService.scala [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+import org.scalamock.scalatest.MockFactory
+import shared.controllers.RequestContext
+import shared.services.ServiceOutcome
+import v1.models.request.createAmend.CreateAmendForeignRequest
+
+import scala.concurrent.{ExecutionContext, Future}
+
+trait MockCreateAmendForeignService extends MockFactory {
+
+  val mockAmendForeignService: CreateAmendForeignService = mock[CreateAmendForeignService]
```
Comment (SamuelJacobsHMRC) [inline test/v1/services/MockCreateAmendForeignService.scala:29]:
```suggestion
  val mockCreateAmendForeignService: CreateAmendForeignService = mock[CreateAmendForeignService]
```

File: app/definition/ApiDefinitionFactory.scala [hunk 9]
```diff
@@ -38,26 +38,26 @@ class ApiDefinitionFactory @Inject() (appConfig: AppConfig) {
...
       api = APIDefinition(
         name = "Individuals Foreign Income (MTD)",
         description = "An API for providing individual foreign income data",
         context = appConfig.apiGatewayContext,
-        categories = Seq("INCOME_TAX_MTD"),
-        versions = Seq(
+        categories = List("INCOME_TAX_MTD"),
+        versions = List(
           APIVersion(
             version = VERSION_1,
```
[the remark this answers is not in the document]
Comment (mattstephens-hmrc) [reply app/definition/ApiDefinitionFactory.scala:62]: Yes, definitely. Not in this PR though :)

**Pull request merged**